force-app/main/default/classes/DMLOperations.cls

```diff
@@ -179,15 +277,22 @@ public with sharing class DMLOperations {
 	 * @param contactsToUpsert List of Contacts to be upserted and linked to accounts.
 	 */
 	public static void upsertAccountsWithContacts(List<Contact> contactsToUpsert) {
-		// Loop through each Contact
-
-		// Extract the Account Name from Contact's LastName
-
-		// Get the account based on the name or create one if it doesn't exist
-
-		// Link the Contact to its Account using the Account Id
-
+		// list to hold the Contacts to upsert
+        List<Contact> contactsToUpsertFinal = new List<Contact>();
+        // Loop through each Contact
+        for (Contact contact : contactsToUpsert) {
+            // Extract the Account Name from Contact's LastName
+            String accountName = contact.LastName;
+            // Get the account based on the name or create one if it doesn't exist
+            Account account = upsertAccount(accountName);
```

Comment:
This causes DML + SOQL within a for loop: find a way to do this without the DML inside of the loop. Using a single query and some maps can tell you if it exists or not and if you need to upsert new accounts before you upsert contacts.